// typed-redux-kit.mapped-reducer/node-build/lib/MappedPipeReducer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
class MappedPipeReducer {
    constructor(opts = {}) {
        this.reducerMap = new Map();
        /**
         * Append reducer functions for the given key
         */
        this.unshift = (actionTypeOrActionTypes, reducerOrReducers) => {
            const actionTypes = util_1.getActionTypes(actionTypeOrActionTypes);
            const reducers = Array.isArray(reducerOrReducers)
                ? reducerOrReducers
                : [reducerOrReducers];
            actionTypes.forEach((actionType) => {
                const reducerArray = this.reducerMap.has(actionType)
                    ? this.reducerMap.get(actionType)
                    : [];
                reducerArray.unshift(...reducers);
                this.reducerMap.set(actionType, reducerArray);
            });
            return this;
        };
        this.prepend = this.unshift;
        /**
         * Append reducer functions for the given key
         */
        this.push = (actionTypeOrActionTypes, reducerOrReducers) => {
            const actionTypes = util_1.getActionTypes(actionTypeOrActionTypes);
            const reducers = Array.isArray(reducerOrReducers)
                ? reducerOrReducers
                : [reducerOrReducers];
            actionTypes.forEach((actionType) => {
                const reducerArray = this.reducerMap.has(actionType)
                    ? this.reducerMap.get(actionType)
                    : [];
                reducerArray.push(...reducers);
                this.reducerMap.set(actionType, reducerArray);
            });
            return this;
        };
        this.append = this.push;
        /**
         * Replace reducer functions for the given key
         */
        this.set = (actionTypeOrActionTypes, reducerOrReducers) => {
            const actionTypes = util_1.getActionTypes(actionTypeOrActionTypes);
            const reducers = Array.isArray(reducerOrReducers)
                ? reducerOrReducers
                : [reducerOrReducers];
            actionTypes.forEach((actionType) => {
                this.reducerMap.set(actionType, reducers);
            });
            return this;
        };
        /**
         * Replace reducer functions for the given key
         */
        this.delete = (actionTypeOrActionTypes) => {
            const actionTypes = util_1.getActionTypes(actionTypeOrActionTypes);
            actionTypes.forEach((actionType) => {
                this.reducerMap.delete(actionType);
            });
            return this;
        };
        this.get = (actionType) => this.reducerMap.get(actionType);
        this.reduce = (state = this.initialState, action) => {
            if (!this.reducerMap.has(action.type))
                return state;
            const reducers = this.reducerMap.get(action.type);
            return reducers.reduce((aState, reducer) => reducer(aState, action), state);
        };
        this.initialState = opts.initialState;
    }
}
exports.MappedPipeReducer = MappedPipeReducer;
