// typed-redux-kit.trackable/node-build/lib/map.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const trackable_1 = require("./trackable");
const util_1 = require("./util");
class TrackableMap extends trackable_1.Trackable {
    constructor(entryIterableOrObject) {
        super();
        if (entryIterableOrObject) {
            const entryIterable = util_1.resolveEntryFromIterable(entryIterableOrObject);
            this.internalMap = new Map();
            for (let [key, value] of entryIterable) {
                value = trackable_1.initializeValue(value, this);
                this.internalMap.set(key, value);
            }
        }
        else {
            this.internalMap = new Map();
        }
    }
    get size() {
        return this.internalMap.size;
    }
    [Symbol.iterator]() {
        return this.internalMap[Symbol.iterator]();
    }
    onChildChange(child) {
        this.markAsChanged();
    }
    has(key) {
        return this.internalMap.has(key);
    }
    get(key, defaultValue) {
        if (!this.internalMap.has(key)) {
            return defaultValue;
        }
        return this.internalMap.get(key);
    }
    set(key, newValue) {
        const previousValue = this.internalMap.get(key);
        if (previousValue !== newValue) {
            this.markAsChanged();
            this.internalMap.set(key, newValue);
            trackable_1.setParentIfTrackable(newValue, this);
        }
        return this;
    }
    delete(key) {
        if (this.internalMap.has(key)) {
            this.markAsChanged();
            this.internalMap.delete(key);
        }
        return this;
    }
    update(key, mutator, defaultValue) {
        const value = this.get(key, defaultValue);
        this.set(key, mutator(value));
        return this;
    }
    merge(partial) {
        const entries = util_1.resolveEntryFromIterable(partial);
        for (const [key, value] of entries) {
            this.set(key, value);
        }
        return this;
    }
    clear() {
        if (this.internalMap.size > 0) {
            this.internalMap = new Map();
            this.markAsChanged();
        }
    }
    entries() {
        return this.internalMap.entries();
    }
    toEntryArray() {
        return util_1.convertIterableToArray(this.entries());
    }
    keys() {
        return this.internalMap.keys();
    }
    toKeyArray() {
        return util_1.convertIterableToArray(this.keys());
    }
    values() {
        return this.internalMap.values();
    }
    toValueArray() {
        return util_1.convertIterableToArray(this.values());
    }
    mapToArray(callback) {
        const array = [];
        for (const [key, value] of this.internalMap.entries()) {
            array.push(callback(value, key));
        }
        return array;
    }
    map(callback) {
        const newMap = new TrackableMap();
        for (const [key, value] of this.internalMap.entries()) {
            newMap.set(key, callback(value, key));
        }
        newMap.$trackable.isChanged = false;
        return newMap;
    }
    filter(callback) {
        const newMap = new TrackableMap();
        for (const [key, value] of this.internalMap.entries()) {
            if (callback(value, key)) {
                newMap.set(key, value);
            }
        }
        newMap.$trackable.isChanged = false;
        return newMap;
    }
    clone() {
        return new TrackableMap(this);
    }
    toJS(shallow = false) {
        const pureObject = {};
        for (const [key, value] of this) {
            pureObject[key] = !shallow && trackable_1.isTrackable(value)
                ? value.toJS()
                : value;
        }
        return pureObject;
    }
}
exports.default = TrackableMap;
