// typed-redux-kit.trackable/node-build/lib/record.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const trackable_1 = require("./trackable");
const util_1 = require("./util");
exports.TrackableRecord = (defaultValue) => {
    class ExtendedTrackableRecord extends TrackableRecordClass {
        clone() {
            return new ExtendedTrackableRecord(Object.assign({}, defaultValue, this.internalObject));
        }
    }
    const keys = Object.keys(defaultValue);
    keys.forEach(key => {
        Object.defineProperty(ExtendedTrackableRecord.prototype, key, {
            set(newValue) {
                this.set(key, newValue);
            },
            get() {
                return this.get(key);
            },
        });
    });
    return (object) => {
        const record = new ExtendedTrackableRecord(Object.assign({}, defaultValue, object));
        return record;
    };
};
class TrackableRecordClass extends trackable_1.Trackable {
    constructor(entryIterableOrObject) {
        super();
        if (entryIterableOrObject) {
            const entryIterable = util_1.resolveEntryFromIterable(entryIterableOrObject);
            this.internalObject = {};
            for (let [key, value] of entryIterable) {
                value = trackable_1.initializeValue(value, this);
                this.internalObject[key] = value;
            }
        }
        else {
            this.internalObject = {};
        }
    }
    [Symbol.iterator]() {
        const entries = util_1.getEntries(this.internalObject);
        return entries[Symbol.iterator]();
    }
    onChildChange(child) {
        this.markAsChanged();
    }
    get(key, defaultValue) {
        if (this.internalObject[key] === undefined)
            return defaultValue;
        return this.internalObject[key];
    }
    set(key, newValue) {
        const previousValue = this.internalObject[key];
        if (previousValue !== newValue) {
            this.markAsChanged();
            this.internalObject[key] = newValue;
            if (trackable_1.isTrackable(newValue)) {
                trackable_1.setParentIfTrackable(newValue, this);
            }
        }
        return this;
    }
    delete(key) {
        if (this.internalObject[key] !== undefined) {
            this.markAsChanged();
            this.internalObject[key] = undefined;
        }
        return this;
    }
    update(key, mutator) {
        const value = this.get(key);
        this.set(key, mutator(value));
        return this;
    }
    merge(partial) {
        const entries = util_1.resolveEntryFromIterable(partial);
        for (const [key, value] of entries) {
            this.set(key, value);
        }
        return this;
    }
    clone() {
        return new TrackableRecordClass(this.internalObject);
    }
    toJS(shallow = false) {
        const pureObject = {};
        for (const [key, value] of this) {
            pureObject[key] = !shallow && trackable_1.isTrackable(value)
                ? value.toJS()
                : value;
        }
        return pureObject;
    }
}
exports.TrackableRecordClass = TrackableRecordClass;
exports.default = exports.TrackableRecord;
