// typed-redux-kit.trackable/node-build/lib/trackable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTrackable = (maybeTrackable) => {
    if (maybeTrackable == null)
        return false;
    return !!maybeTrackable.$trackable;
};
exports.setParentIfTrackable = (maybeTrackable, parent) => {
    if (exports.isTrackable(maybeTrackable)) {
        maybeTrackable.setParent(parent);
    }
};
exports.initializeValue = (maybeTrackable, parent) => {
    if (exports.isTrackable(maybeTrackable)) {
        if (maybeTrackable.$trackable.isChanged) {
            maybeTrackable = maybeTrackable.clone();
        }
        maybeTrackable.setParent(parent);
    }
    return maybeTrackable;
};
class Trackable {
    constructor() {
        this.$trackable = {
            isChanged: false,
            shallow: false,
        };
    }
    markAsChanged() {
        this.$trackable.isChanged = true;
        if (this.$trackable.parent) {
            this.$trackable.parent.onChildChange(this);
        }
    }
    setParent(parent) {
        this.$trackable.parent = parent;
    }
}
exports.Trackable = Trackable;
