// typed-redux-kit.trackable/node-build/lib/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEntries = Object.entries;
exports.getValues = Object.values;
exports.resolveEntryFromIterable = (entryIterableOrObject) => {
    return !!entryIterableOrObject[Symbol.iterator]
        ? entryIterableOrObject
        : Object.entries(entryIterableOrObject);
};
exports.convertIterableToArray = (iterable) => {
    const array = [];
    for (const item of iterable) {
        array.push(item);
    }
    return array;
};
function isPlainObject(value) {
    if (value === null || typeof value !== 'object')
        return false;
    const proto = Object.getPrototypeOf(value);
    return proto === Object.prototype || proto === null;
}
exports.isPlainObject = isPlainObject;
function getActionTypes(actionTypeOrActionTypes) {
    return Array.isArray(actionTypeOrActionTypes)
        ? actionTypeOrActionTypes
        : isPlainObject(actionTypeOrActionTypes)
            ? exports.getValues(actionTypeOrActionTypes)
            : [actionTypeOrActionTypes];
}
exports.getActionTypes = getActionTypes;
