// typed-redux-kit.trackable/build/lib/map.js
import { Trackable, isTrackable, setParentIfTrackable, initializeValue, } from './trackable';
import { resolveEntryFromIterable, convertIterableToArray, } from './util';
class TrackableMap extends Trackable {
    constructor(entryIterableOrObject) {
        super();
        if (entryIterableOrObject) {
            const entryIterable = resolveEntryFromIterable(entryIterableOrObject);
            this.internalMap = new Map();
            for (let [key, value] of entryIterable) {
                value = initializeValue(value, this);
                this.internalMap.set(key, value);
            }
        }
        else {
            this.internalMap = new Map();
        }
    }
    get size() {
        return this.internalMap.size;
    }
    [Symbol.iterator]() {
        return this.internalMap[Symbol.iterator]();
    }
    onChildChange(child) {
        this.markAsChanged();
    }
    has(key) {
        return this.internalMap.has(key);
    }
    get(key, defaultValue) {
        if (!this.internalMap.has(key)) {
            return defaultValue;
        }
        return this.internalMap.get(key);
    }
    set(key, newValue) {
        const previousValue = this.internalMap.get(key);
        if (previousValue !== newValue) {
            this.markAsChanged();
            this.internalMap.set(key, newValue);
            setParentIfTrackable(newValue, this);
        }
        return this;
    }
    delete(key) {
        if (this.internalMap.has(key)) {
            this.markAsChanged();
            this.internalMap.delete(key);
        }
        return this;
    }
    update(key, mutator, defaultValue) {
        const value = this.get(key, defaultValue);
        this.set(key, mutator(value));
        return this;
    }
    merge(partial) {
        const entries = resolveEntryFromIterable(partial);
        for (const [key, value] of entries) {
            this.set(key, value);
        }
        return this;
    }
    clear() {
        if (this.internalMap.size > 0) {
            this.internalMap = new Map();
            this.markAsChanged();
        }
    }
    entries() {
        return this.internalMap.entries();
    }
    toEntryArray() {
        return convertIterableToArray(this.entries());
    }
    keys() {
        return this.internalMap.keys();
    }
    toKeyArray() {
        return convertIterableToArray(this.keys());
    }
    values() {
        return this.internalMap.values();
    }
    toValueArray() {
        return convertIterableToArray(this.values());
    }
    mapToArray(callback) {
        const array = [];
        for (const [key, value] of this.internalMap.entries()) {
            array.push(callback(value, key));
        }
        return array;
    }
    map(callback) {
        const newMap = new TrackableMap();
        for (const [key, value] of this.internalMap.entries()) {
            newMap.set(key, callback(value, key));
        }
        newMap.$trackable.isChanged = false;
        return newMap;
    }
    filter(callback) {
        const newMap = new TrackableMap();
        for (const [key, value] of this.internalMap.entries()) {
            if (callback(value, key)) {
                newMap.set(key, value);
            }
        }
        newMap.$trackable.isChanged = false;
        return newMap;
    }
    clone() {
        return new TrackableMap(this);
    }
    toJS(shallow = false) {
        const pureObject = {};
        for (const [key, value] of this) {
            pureObject[key] = !shallow && isTrackable(value)
                ? value.toJS()
                : value;
        }
        return pureObject;
    }
}
export default TrackableMap;
