// typed-redux-kit.trackable/build/lib/record.js
import { Trackable, isTrackable, setParentIfTrackable, initializeValue, } from './trackable';
import { resolveEntryFromIterable, getEntries, } from './util';
export const TrackableRecord = (defaultValue) => {
    class ExtendedTrackableRecord extends TrackableRecordClass {
        clone() {
            return new ExtendedTrackableRecord(Object.assign({}, defaultValue, this.internalObject));
        }
    }
    const keys = Object.keys(defaultValue);
    keys.forEach(key => {
        Object.defineProperty(ExtendedTrackableRecord.prototype, key, {
            set(newValue) {
                this.set(key, newValue);
            },
            get() {
                return this.get(key);
            },
        });
    });
    return (object) => {
        const record = new ExtendedTrackableRecord(Object.assign({}, defaultValue, object));
        return record;
    };
};
export class TrackableRecordClass extends Trackable {
    constructor(entryIterableOrObject) {
        super();
        if (entryIterableOrObject) {
            const entryIterable = resolveEntryFromIterable(entryIterableOrObject);
            this.internalObject = {};
            for (let [key, value] of entryIterable) {
                value = initializeValue(value, this);
                this.internalObject[key] = value;
            }
        }
        else {
            this.internalObject = {};
        }
    }
    [Symbol.iterator]() {
        const entries = getEntries(this.internalObject);
        return entries[Symbol.iterator]();
    }
    onChildChange(child) {
        this.markAsChanged();
    }
    get(key, defaultValue) {
        if (this.internalObject[key] === undefined)
            return defaultValue;
        return this.internalObject[key];
    }
    set(key, newValue) {
        const previousValue = this.internalObject[key];
        if (previousValue !== newValue) {
            this.markAsChanged();
            this.internalObject[key] = newValue;
            if (isTrackable(newValue)) {
                setParentIfTrackable(newValue, this);
            }
        }
        return this;
    }
    delete(key) {
        if (this.internalObject[key] !== undefined) {
            this.markAsChanged();
            this.internalObject[key] = undefined;
        }
        return this;
    }
    update(key, mutator) {
        const value = this.get(key);
        this.set(key, mutator(value));
        return this;
    }
    merge(partial) {
        const entries = resolveEntryFromIterable(partial);
        for (const [key, value] of entries) {
            this.set(key, value);
        }
        return this;
    }
    clone() {
        return new TrackableRecordClass(this.internalObject);
    }
    toJS(shallow = false) {
        const pureObject = {};
        for (const [key, value] of this) {
            pureObject[key] = !shallow && isTrackable(value)
                ? value.toJS()
                : value;
        }
        return pureObject;
    }
}
export default TrackableRecord;
